File: trunk/PhotSpot/src/com/erdao/PhotSpot/GeoClusterer.java
/*
 * Copyright (C) 2009 Huan Erdao
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *	  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.erdao.PhotSpot;

import java.util.ArrayList;
import java.util.List;
import java.util.Timer;
import java.util.TimerTask;

import android.content.Context;
import android.graphics.Point;
import android.os.Handler;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.TextView;

import com.google.android.maps.GeoPoint;
import com.google.android.maps.MapController;
import com.google.android.maps.MapView;
import com.google.android.maps.Overlay;
import com.google.android.maps.Projection;

/* Class for Clustering geotagged content 
 * this clustering came from "markerclusterer" which is available as opensource at
 * http://code.google.com/p/gmaps-utility-library/
 * this is android ported version with modification to fit to the application
 */
public class GeoClusterer {
	
	/* variables */
	private final Context context_;
	private final MapView mapView_;
	private final FrameLayout imageFrame_;
	private ArrayList<PhotoItem> items_ = new ArrayList<PhotoItem>();
	private ArrayList<PhotoItem> leftItems_ = new ArrayList<PhotoItem>();
	private ArrayList<GeoCluster> clusters_ = new ArrayList<GeoCluster>();
	private int gridSize_ = 40;
	private GeoCluster selcluster_ = null;
	private int checkcnt_ = 0;
	private GeoBounds checkBound_;
	private boolean isMoving_;
	private Handler handler_;

	/* constructor */
	public GeoClusterer(Context context, MapView mapView, FrameLayout imageFrame){
		context_ = context;
		mapView_ = mapView;
		imageFrame_ = imageFrame;
		handler_ = new Handler();
	}

	/* addItem */
	public void addItem(PhotoItem item) {
		// if not in viewport, add to leftItems_
		if(!isItemInViewport(item)) {
			leftItems_.add(item);
			return;
		}
		// else add to items_;
		items_.add(item);
		int length = clusters_.size();
		GeoCluster cluster = null;
		Projection proj = mapView_.getProjection();
		Point pos = proj.toPixels(item.getLocation(), null);
		// check existing cluster
		for(int i=length-1; i>=0; i--) {
			  cluster = clusters_.get(i);
			  GeoPoint gpCenter = cluster.getLocation();
			  if(gpCenter == null)
				  continue;
			  Point ptCenter = proj.toPixels(gpCenter,null);
			  // find a cluster which contains the marker.
			  if(pos.x >= ptCenter.x - gridSize_ && pos.x <= ptCenter.x + gridSize_ &&
				  pos.y >= ptCenter.y - gridSize_ && pos.y <= ptCenter.y + gridSize_) {
				  cluster.addItem(item);
//				  cluster.redraw();
				  return;
			  }
		}
		// No cluster contain the marker, create a new cluster.
		cluster = new GeoCluster(this);
		cluster.addItem(item);
		// Add Cluster to ArrayList
		clusters_.add(cluster);
//		cluster.redraw();
	}

	/* redraw */
	public void redraw(){
		for(int i=0; i<clusters_.size(); i++) {
			clusters_.get(i).redraw();
		}
	}

	/* zoomInFixing */
	public void zoomInFixing(){
		if(selcluster_!=null){
			GeoPoint gpt = selcluster_.getSelectedItemLocation();
			if(getCurBounds().isInBounds(gpt)){
				Projection pro = mapView_.getProjection();
				Point ppt = pro.toPixels(gpt, null);
				MapController mapCtrl = mapView_.getController();
				mapCtrl.zoomInFixing(ppt.x, ppt.y);
			}
			else{
				MapController mapCtrl = mapView_.getController();
				mapCtrl.zoomIn();
			}
		}
		else{
			MapController mapCtrl = mapView_.getController();
			mapCtrl.zoomIn();
		}
	}
	
	/* isItemInViewport */
	private boolean isItemInViewport(PhotoItem item){
		checkBound_ = getCurBounds();
		isMoving_ = false;
		return checkBound_.isInBounds(item.getLocation());
	}

	/* getCurBounds */
	private GeoBounds getCurBounds(){
		Projection proj = mapView_.getProjection();
		return new GeoBounds(proj.fromPixels(0,0),proj.fromPixels(mapView_.getWidth(),mapView_.getHeight()));
	}

	/* getMapView */
	private MapView getMapView(){
		return mapView_;
	}

	/* getGridSize */
	private int getGridSize(){
		return gridSize_;
	}

	/* getClustersInViewport */
	private ArrayList<GeoCluster> getClustersInViewport() {
		GeoBounds curBounds = getCurBounds();
		ArrayList<GeoCluster> clusters = new ArrayList<GeoCluster>();
		for(int i=0; i<clusters_.size(); i++) {
			GeoCluster cluster = clusters_.get(i);
			  if(cluster.isInBounds(curBounds)) {
				  clusters.add(cluster);
			  }
		}
		return clusters;
	}

	/* addLeftItems */
	private void addLeftItems() {
		if(leftItems_.size()==0){
			return;
		}
		ArrayList<PhotoItem> leftItems = new ArrayList<PhotoItem>();
		leftItems.addAll(leftItems_);
		leftItems_.clear();
		for(int i=0; i<leftItems.size(); i++) {
			addItem(leftItems.get(i));
		}
	}

	/* reAddItems */
	private void reAddItems(ArrayList<PhotoItem> items) {
		int len = items.size();
		for(int i=len-1; i>=0; i--) {
			addItem(items.get(i));
		}
		addLeftItems();
	}

	/* resetViewport */
	private void resetViewport() {
//		Log.i("DEBUG","resetViewport");
		ArrayList<GeoCluster> clusters = getClustersInViewport();
		ArrayList<PhotoItem> tmpItems = new ArrayList<PhotoItem>();
		int removed = 0;
		for(int i=0; i<clusters.size(); i++) {
			GeoCluster cluster = clusters.get(i);
			int oldZoom = cluster.getCurrentZoom();
			int curZoom = mapView_.getZoomLevel();
			// If the cluster zoom level changed then destroy the cluster and collect its markers.
			if(curZoom != oldZoom) {
				tmpItems.addAll(cluster.getItems());
				cluster.clearMarker();
				removed++;
				for(int j=0; j<clusters_.size(); j++) {
					if(cluster == clusters_.get(j)) {
						clusters_.remove(j);
					}
				}
			}
		}
		reAddItems(tmpItems);
		redraw();
		// Add the markers collected into marker cluster to reset
		if(removed>0){
			boolean dismiss = true;
			for(int i=0; i<clusters_.size(); i++) {
				GeoCluster cluster = clusters_.get(i);
				if( cluster.isSelected() ){
					dismiss = false;
					selcluster_ = cluster;
					cluster.showGallery();
					MapController mapCtrl = mapView_.getController();
					mapCtrl.animateTo(cluster.getLocation());
					break;
				}
			}
			if(dismiss){
				imageFrame_.setVisibility(View.GONE);
				TextView txtView = (TextView)imageFrame_.findViewById(R.id.copyright);
				txtView.setVisibility(View.GONE);
			}
		}
		mapView_.invalidate();
	}

	/* onNotifyDraw
	 * Little tweek to enable catching zoom/move event when markers are present.
	 * hope there will be event notification for android equivalent to javascriptin the future....
	 */
	public void onNotifyDraw(){
		if(!isMoving_){
			GeoBounds curBnd = getCurBounds();
			if( !checkBound_.isEqual(curBnd) ){
				isMoving_ = true;
				checkBound_ = curBnd;
//				Log.i("DEBUG","movestart");
				Timer timer = new Timer(true);
				timer.schedule(
					new TimerTask() {
						public void run() {
							//Log.i("DEBUG","timertask");
							GeoBounds curBnd = getCurBounds();
							if( checkBound_.isEqual(curBnd) ){
								isMoving_ = false;
								//Log.i("DEBUG","moveend");
								this.cancel();
								handler_.post( new Runnable() {
									public void run() {
										resetViewport();
									}
								});
							}
							checkBound_ = curBnd;
						}
					}, 500, 500
				);
			}
		}
	}

	/* onTap, not usual onTap method. called from Cluster Layer */
	public void onTap(GeoCluster caller,boolean tapped) {
		if(tapped){
			if(selcluster_ == caller)
				return;
			for(int i=0; i<clusters_.size(); i++) {
				if(selcluster_==clusters_.get(i)) {
					clusters_.get(i).clearSelect();
				}
			}
			selcluster_ = caller;
		}
		else{
			checkcnt_++;
			if( checkcnt_ == clusters_.size() ){
				checkcnt_ = 0;
				/* TODO : do something if no marker was tapped
				 */
//				selcluster_.clearSelect();
			}
		}
	}

	/* getCluster */
	public GeoCluster getCluster(int id) {
		return clusters_.get(id);
	}
	
	/* Cluster class - GeoCluster
	 * contains single marker object(ClusterMarker). mostly wraps methods in ClusterMarker.
	 */
	public class GeoCluster {
		/* variables */
		private final GeoClusterer clusterer_;
		private final MapView mapView_;
		private GeoPoint center_;
		private ArrayList<PhotoItem> items_ = new ArrayList<PhotoItem>();
		private ClusterMarker clusterMarker_;
		private int zoom_;

		/* constructor */
		public GeoCluster(GeoClusterer clusterer){
			clusterer_ = clusterer;
			mapView_ = clusterer_.getMapView();
			zoom_ = mapView_.getZoomLevel();
		}
		
		/* addItem */
		public void addItem(PhotoItem item){
			if(center_ == null){
				center_ = item.getLocation();
			}
			items_.add(item);
		}
		
		/* getLocation */
		public GeoPoint getLocation(){
			return center_;
		}

		/* getSelectedItemLocation */
		public GeoPoint getSelectedItemLocation(){
			return clusterMarker_.getSelectedItemLocation();
		}

		/* showGallery */
		public void showGallery(){
			clusterMarker_.showGallery();
		}

		/* clearSelect */
		public void clearSelect(){
			clusterMarker_.clearSelect();
		}

		/* isSelected */
		public boolean isSelected(){
			return clusterMarker_.isSelected();
		}

		/* getCurrentZoom */
		public int getCurrentZoom(){
			return zoom_;
		}
		
		/* getItems */
		public ArrayList<PhotoItem> getItems(){
			return items_;
		}

		/* onNotifyDraw */
		public void onNotifyDraw(){
			clusterer_.onNotifyDraw();
		}

		/* onTap - not usual onTap. called from ClusterMarker */
		public void onTap(boolean flg) {
			clusterer_.onTap(this,flg);
		}
		
		/* clearMarker */
		public void clearMarker() {
			if(clusterMarker_ != null) {
				List<Overlay> mapOverlays = mapView_.getOverlays();
				if(mapOverlays.contains(clusterMarker_)){
					mapOverlays.remove(clusterMarker_);
				}
			}
			items_ = null;
		}

		/* redraw */
		public void redraw(){
			if(!isInBounds(clusterer_.getCurBounds())) {
				return;
			}
			if(clusterMarker_ == null) {
				clusterMarker_ = new ClusterMarker(this,mapView_,context_,imageFrame_);
				List<Overlay> mapOverlays = mapView_.getOverlays();
				mapOverlays.add(clusterMarker_);
			}
		}

		/* isInBounds */
		private boolean isInBounds(GeoBounds bounds) {
			if(center_ == null) {
				return false;
			}
			Projection pro = mapView_.getProjection();
			Point nw = pro.toPixels(bounds.getNorthWest(),null);
			Point se = pro.toPixels(bounds.getSouthEast(),null);
			Point centxy = pro.toPixels(center_,null);
			boolean inViewport = true;
			int gridSize = clusterer_.getGridSize();
			if(zoom_ != mapView_.getZoomLevel()) {
				int diff = mapView_.getZoomLevel() - zoom_;
				gridSize = (int) (Math.pow(2, diff) * gridSize);
			}
			if(nw.x != se.x && (centxy.x + gridSize < nw.x || centxy.x - gridSize > se.x)) {
				inViewport = false;
			}
			if(inViewport && (centxy.y + gridSize < nw.y || centxy.y - gridSize > se.y)) {
				inViewport = false;
			}
			return inViewport;
		}
	};
}

File: trunk/markerclusterer_for_v3_api/geoclusterer.js
/**
 * @name GeoClusterer
 * @version 1.0
 * @author Huan Erdao
 * @copyright (c) 2009 Huan Erdao
 * @fileoverview
 * This javascript library comes from markerclusterer by Xiaoxi Wu
 * which partially ported to work with google maps v3 apis.
 */

/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Projection overlay helper class
 * came from : http://snippets.aktagon.com/snippets/377-How-to-find-a-Google-Map-marker-s-exact-position
 * @constructor
 * @param {google.maps.Map} map The map that the markers should be added to.
 */
function ProjectionHelperOverlay(map) {
	this.setMap(map);
}
ProjectionHelperOverlay.prototype = new google.maps.OverlayView();
ProjectionHelperOverlay.prototype.draw = function() {
	if (!this.ready){
		this.ready = true;
		google.maps.event.trigger(this, 'ready');
	}
};

/**
 * GeoCluster class
 * @constructor
 * @param {google.maps.Map} map The map that the markers should be added to.
 */
function GeoClusterer(map) {
	// private members
	var clusters_ = [];
	var map_ = map;
	var maxZoom_ = null;
	var me_ = this;
	var gridSize_ = 90;
	var leftItems_ = [];
	var mcfn_ = null;
	var projectionHelper_ = new ProjectionHelperOverlay(map_);

	/**
	 * addItem.
	 */
	this.addItem = function (item) {
		if (!isItemInViewport_(item)) {
			leftItems_.push(item);
			return;
		}
		var projection = projectionHelper_.getProjection();
		var pos = projection.fromLatLngToDivPixel(item.latlng);
		var length = clusters_.length;
		var cluster = null;
		for (var i = length - 1; i >= 0; i--) {
			cluster = clusters_[i];
			var gpCenter = cluster.getCenter();
			if (gpCenter === null) {
				continue;
			}
			var ptCenter = projection.fromLatLngToDivPixel(gpCenter);
			// Found a cluster which contains the marker.
			if (pos.x >= ptCenter.x - gridSize_ && pos.x <= ptCenter.x + gridSize_ &&
							pos.y >= ptCenter.y - gridSize_ && pos.y <= ptCenter.y + gridSize_) {
				cluster.addItem(item);
				return;
			}
	    }
		// No cluster contain the marker, create a new cluster.
		createCluster(item);
	};
	
	/**
	 * Create Cluster Object.
	 * override this method, if you want to use custom GeoCluster class.
	 * @param item GeoItem to be set to cluster.
	 */
	 function createCluster(item){
	 	 cluster = new Cluster(me_, map);
		 cluster.addItem(item);
		 clusters_.push(cluster);
	 }
	 
	/**
	 * Redraw all clusters in viewport.
	 */
	this.redraw_ = function () {
		for (var i = 0; i < clusters_.length; ++i) {
			clusters_[i].redraw_(true);
		}
	};
	
	this.getProjectionHelper = function() {
		return projectionHelper_;
	};
	
	/**
	 * Get all clusters in viewport.
	 */
	this.getClustersInViewport_ = function () {
		var clusters = [];
		var curBounds = map_.getBounds();
		for (var i = 0; i < clusters_.length; i ++) {
			if (clusters_[i].isInBounds(curBounds)) {
				clusters.push(clusters_[i]);
			}
		}
		return clusters;
	};

	/**
	 * addleftItems_
	 */
	function addleftItems_() {
		if (leftItems_.length === 0) {
			return;
		}
		var leftItems = leftItems_.slice(0);
		leftItems_.length = 0;
		for (i = 0; i < leftItems.length; ++i) {
			me_.addItem(leftItems[i]);
		}
	}

	/**
	 * reAddItems_
	 */
	function reAddItems_(items) {
		var len = items.length;
		for (var i = len - 1; i >= 0; --i) {
			me_.addItem(items[i]);
		}
		addleftItems_();
	}

	/**
	 * Remove all markers from MarkerClusterer.
	 */
	this.clear = function () {
		for (var i = 0; i < clusters_.length; ++i) {
			if (typeof clusters_[i] !== "undefined" && clusters_[i] !== null) {
				clusters_[i].clear();
			}
		}
		clusters_ = [];
		leftItems_ = [];
		if(mcfn_){
			google.maps.event.removeListener(mcfn_);
			mcfn_ = null;
		}
	};

	/**
	 * addItems
	 */
	this.addItems = function (items) {
		for (var i = 0; i < items.length; ++i) {
			this.addItem(items[i]);
		}
		this.redraw_();
		// when map move end, regroup.
		mcfn_ = google.maps.event.addListener(map_, "bounds_changed", function () {
			me_.resetViewport();
		});
	};

	/**
	 * isItemInViewport_
	 */
	function isItemInViewport_(item) {
		return map_.getBounds().contains(item.latlng);
	}

	/**
	 * Get grid size
	 */
	this.getGridSize_ = function () {
		return gridSize_;
	};

	/**
	 * Collect all markers of clusters in viewport and regroup them.
	*/
	this.resetViewport = function () {
		var clusters = this.getClustersInViewport_();
		var tmpItems = [];
		var removed = 0;
		for (var i = 0; i < clusters.length; ++i) {
			var cluster = clusters[i];
			var oldZoom = cluster.getCurrentZoom();
			if (oldZoom === null) {
				continue;
			}
			var curZoom = map_.getZoom();
			if (curZoom !== oldZoom) {
				// If the cluster zoom level changed then destroy the cluster
				// and collect its markers.
				var citms = cluster.getItems();
				for (j = 0; j < citms.length; ++j) {
					tmpItems.push(citms[j]);
				}
				cluster.clear();
				removed++;
				for (j = 0; j < clusters_.length; ++j) {
					if (cluster === clusters_[j]) {
						clusters_.splice(j, 1);
					}
				}
			}
		}
		// Add the markers collected into marker cluster to reset
		reAddItems_(tmpItems);
		this.redraw_();
	};

	this.onClickMarker = function (cluster){
	};

}

/**
 Cluster 
 */
function Cluster(geoClusterer,map) {
	var center_ = null;
	var items_ = [];
	var geoClusterer_ = geoClusterer;
	var map_ = map;
	var clusterMarker_ = null;
	var zoom_ = map_.getZoom();

	this.getItems = function () {
		return items_;
	};

	this.isInBounds = function (bounds) {
		if (center_ === null) {
			return false;
		}
		if (!bounds) {
			bounds = map_.getBounds();
		}
		var ph = geoClusterer_.getProjectionHelper();
		var projection = ph.getProjection();
		var sw = projection.fromLatLngToDivPixel(bounds.getSouthWest());
		var ne = projection.fromLatLngToDivPixel(bounds.getNorthEast());
		var mapcenter = map_.getCenter();
		var centerxy = projection.fromLatLngToDivPixel(center_);
		var inViewport = true;
		var gridSize = geoClusterer_.getGridSize_();
		if (zoom_ !== map_.getZoom()) {
			var dl = map_.getZoom() - zoom_;
			gridSize = Math.pow(2, dl) * gridSize;
		}
		if (ne.x !== sw.x && (centerxy.x + gridSize < sw.x || centerxy.x - gridSize > ne.x)) {
			inViewport = false;
		}
		if (inViewport && (centerxy.y + gridSize < ne.y || centerxy.y - gridSize > sw.y)) {
			inViewport = false;
		}
		return inViewport;
	};

	/**
	 * Get cluster center.
	 */
	this.getCenter = function () {
		return center_;
	};

	/**
	 * Add a marker.
	 */
	this.addItem = function (item) {
		if (center_ === null) {
			center_ = item.latlng;
		}
		items_.push(item);
	};

	/**
	 * Get current zoom level of this cluster.
	 */
	this.getCurrentZoom = function () {
		return zoom_;
	};

	this.redraw_ = function(){
		if(!this.isInBounds()) {
			return;
		}
		if(clusterMarker_ == null) {
			clusterMarker_ = new ClusterMarker_(map_,this,geoClusterer_.onClickMarker);
		}
	}

	/**
	 * Remove all the markers from this cluster.
	*/
	this.clear = function () {
		if (clusterMarker_ !== null) {
			clusterMarker_.setMap(null);
			clusterMarker_ = null;
		}
		items_ = [];
	};

	/**
	 * Get number of markers.
	 */
	this.getItemSize = function () {
		return items_.length;
	};
}

/**
 * ClusterMarker_ creates a marker that shows the number of markers that
 */
function ClusterMarker_(map, cluster, clickCallback) {
	this.cluster_ = cluster;
	this.latlng_ = cluster.getCenter();
	var count = cluster.getItemSize();
	this.count_ = count;
	this.icon_ = count >= 10 ? "http://gmaps-utility-library.googlecode.com/svn/trunk/markerclusterer/images/m3.png" : "http://gmaps-utility-library.googlecode.com/svn/trunk/markerclusterer/images/m1.png";
	this.width_ = count >= 10 ? 66 : 53;
	this.height_ = count >= 10 ? 65 : 52;
	this.mstyle_ = "background-image:url(" + this.icon_ + ");";
	this.mstyle_ += 'height:' + this.height_ + 'px;line-height:' + this.height_ + 'px;';
	this.mstyle_ += 'width:' + this.width_ + 'px;text-align:center;';
	this.clickCallback_ = clickCallback;
	this.setMap(map);
}

ClusterMarker_.prototype = new google.maps.OverlayView();

ClusterMarker_.prototype.draw = function() {
	this.projection_ = this.getProjection();
	var map = this.getMap();
	var pos = this.projection_.fromLatLngToDivPixel(this.latlng_);
	pos.x -= parseInt(this.width_ / 2, 10);
	pos.y -= parseInt(this.height_ / 2, 10);
	this.div_.style.left = pos.x + "px";
	this.div_.style.top = pos.y + "px";
	var txtColor = 'white';
	this.div_.style.cssText = this.mstyle_ + 'cursor:pointer; top:' + pos.y + "px;left:"
		+ pos.x + "px; color:" + txtColor +  ";position:absolute;font-size:11px;"
		+ 'font-family:Arial,sans-serif;font-weight:bold';
	this.div_.innerHTML = this.count_;
};

ClusterMarker_.prototype.onAdd = function() {
	var map = this.getMap();
	var div = document.createElement("div");
	div.style.visibility = 'visible';
	var cluster = this.cluster_;
	var callback = this.clickCallback_;
	google.maps.event.addDomListener(div, "click", function () {
		callback(cluster);
	});
	this.div_ = div;
	this.getPanes().mapPane.appendChild(this.div_);
};

ClusterMarker_.prototype.remove = function() {
	this.div_.parentNode.removeChild(this.div_);
};

